src/core/tsi/s2a/record_protocol/s2a_crypter.cc

```diff
@@ -256,3 +248,244 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+ *  - half_connection: an instance of s2a_half_connection.
+ *  - out_bytes: a pointer to a length 8 array of bytes. **/
+static void sequence_to_bytes(const s2a_half_connection* half_connection,
+                              uint8_t* out_bytes) {
+  GPR_ASSERT(half_connection != nullptr);
+  GPR_ASSERT(out_bytes != nullptr);
+  out_bytes[0] = half_connection->sequence >> 56;
+  out_bytes[1] = half_connection->sequence >> 48;
+  out_bytes[2] = half_connection->sequence >> 40;
+  out_bytes[3] = half_connection->sequence >> 32;
+  out_bytes[4] = half_connection->sequence >> 24;
+  out_bytes[5] = half_connection->sequence >> 16;
+  out_bytes[6] = half_connection->sequence >> 8;
+  out_bytes[7] = half_connection->sequence;
+}
+
+size_t s2a_max_record_overhead(const s2a_crypter* crypter) {
+  GPR_ASSERT(crypter != nullptr);
+  GPR_ASSERT(crypter->out_connection != nullptr);
+  GPR_ASSERT(crypter->out_connection->initialized);
+  return SSL3_RT_HEADER_LENGTH + s2a_tag_size(crypter) + /** record type **/ 1;
+}
+
+static grpc_status_code s2a_write_tls13_record_header(uint8_t record_type,
+                                                      size_t header_size,
+                                                      size_t payload_size,
+                                                      uint8_t* record_header,
+                                                      char** error_details) {
+  GPR_ASSERT(record_header != nullptr);
+  if (header_size != SSL3_RT_HEADER_LENGTH) {
+    *error_details = gpr_strdup(
+        "The header size does not match the size of a TLS 1.3 record header.");
+    return GRPC_STATUS_FAILED_PRECONDITION;
+  }
+  record_header[0] = record_type;
+  const uint16_t wire_version = static_cast<uint16_t>(TLS1_2_VERSION);
+  record_header[1] = wire_version >> 8;
+  record_header[2] = wire_version & 0xff;
+  record_header[3] = payload_size >> 8;
+  record_header[4] = payload_size & 0xff;
+  return GRPC_STATUS_OK;
+}
+
+/** This method returns a buffer of additional data bytes used for
+ *  authenticating a TLS 1.3 record; the caller must free the buffer using
+ *  gpr_free.
+ *  - sequence: a buffer containing the sequence number of a half connection.
+ *  - sequence_size: the size (in bytes) of the |sequence| buffer.
+ *  - record_header: a buffer containing the header of a TLS 1.3 record.
+ *  - header_size: the size (in bytes) of the |record_header| buffer.
+ *  - payload_size: the size (in bytes) of the payload of a TLS 1.3 record. **/
+static uint8_t* s2a_additional_data(uint8_t* sequence, size_t sequence_size,
+                                    uint8_t* record_header, size_t header_size,
+                                    size_t payload_size) {
+  size_t additional_data_size = sequence_size + header_size;
+  GPR_ASSERT(additional_data_size == kTlsAdditionalDataBytesSize);
+  uint8_t* additional_data =
+      (uint8_t*)gpr_malloc(additional_data_size * sizeof(uint8_t));
```

Comment:
Can you use reinterpret_cast<uint*>(...)?

Reply:
Or static_cast ...

Reply:
Changed (and the other similar instances of this).